File: tests/domain/smoothmanifolds/test_Grassmannian.cpp
#include <Lielab.hpp>

#include <catch2/catch_all.hpp>

#include <cmath>
#include <numbers>

TEST_CASE("Grassmannian to_string", "[domain]")
{
    using Lielab::domain::Grassmannian;

    const Grassmannian xblank = Grassmannian();
    CHECK(xblank.to_string() == "Grassmannian(0, 0, R)");

    const Grassmannian x0 = Grassmannian(0, 0);
    CHECK(x0.to_string() == "Grassmannian(0, 0, R)");
    const Grassmannian x1 = Grassmannian(1, 1);
    CHECK(x1.to_string() == "Grassmannian(1, 1, R)");
    const Grassmannian x10 = Grassmannian(5, 10);
    CHECK(x10.to_string() == "Grassmannian(5, 10, R)");
}

TEST_CASE("Grassmannian main_initializer", "[domain]")
{
    using Lielab::domain::Grassmannian;

    const Grassmannian xblank = Grassmannian();
    CHECK(xblank.get_dimension() == 0);

    const Grassmannian x0 = Grassmannian(0, 0);
    CHECK(x0.get_dimension() == 0);
    const Grassmannian x1 = Grassmannian(1, 1);
    CHECK(x1.get_dimension() == 1);
    const Grassmannian x10 = Grassmannian(5, 10);
    CHECK(x10.get_dimension() == 5);

    CHECK_THROWS(Grassmannian(6, 5));
}

TEST_CASE("Grassmannian get_dimension", "[domain]")
{
    using Lielab::domain::Grassmannian;

    Grassmannian zero(0, 0), one(1, 1), two(2, 2), three(3, 3), four(4, 4), five(5, 5), six(6, 6), seven(7, 7), eight(8, 8);

    CHECK(zero.get_dimension() == 0);
    CHECK(one.get_dimension() == 1);
    CHECK(two.get_dimension() == 2);
    CHECK(three.get_dimension() == 3);
    CHECK(four.get_dimension() == 4);
    CHECK(five.get_dimension() == 5);
    CHECK(six.get_dimension() == 6);
    CHECK(seven.get_dimension() == 7);
    CHECK(eight.get_dimension() == 8);
}

TEST_CASE("Grassmannian serialize/unserialize", "[domain]")
{
    /*!
    * Tests the serialize/unserialize operation.
    */

    using Lielab::domain::Grassmannian;

    Grassmannian x0 = Grassmannian(0, 0);
    x0.unserialize({});
    Eigen::VectorXd x0bar = x0.serialize();

    CHECK(x0bar.size() == 0);

    Grassmannian x1 = Grassmannian(1, 2);
    x1.unserialize({1.0, 2.0, 3.0});
    Eigen::VectorXd x1bar = x1.serialize();

    REQUIRE(x1bar.size() == 2);
    CHECK(x1bar(0) == 1.0);
    CHECK(x1bar(1) == 2.0);

    x1.unserialize({4.0, 5.0});
    x1bar = x1.serialize();

    REQUIRE(x1bar.size() == 2);
    CHECK(x1bar(0) == 4.0);
    CHECK(x1bar(1) == 5.0);

    x1.unserialize({6.0});
    x1bar = x1.serialize();

    REQUIRE(x1bar.size() == 2);
    CHECK(x1bar(0) == 6.0);
    CHECK(x1bar(1) == 5.0);

    Grassmannian x2 = Grassmannian(2, 4);
    x2.unserialize({1.0, 2.0, 3.0});
    Eigen::VectorXd x2bar = x2.serialize();

    REQUIRE(x2bar.size() == 4);
    CHECK(x2bar(0) == 1.0);
    CHECK(x2bar(1) == 2.0);
    CHECK(x2bar(2) == 3.0);
    CHECK(x2bar(3) == 0.0);

    x2.unserialize({4.0, 5.0, 6.0, 7.0});
    x2bar = x2.serialize();

    REQUIRE(x2bar.size() == 4);
    CHECK(x2bar(0) == 4.0);
    CHECK(x2bar(1) == 5.0);
    CHECK(x2bar(2) == 6.0);
    CHECK(x2bar(3) == 7.0);

    x2.unserialize({8.0, 9.0, 10.0, 11.0, 12.0});
    x2bar = x2.serialize();

    REQUIRE(x2bar.size() == 4);
    CHECK(x2bar(0) == 8.0);
    CHECK(x2bar(1) == 9.0);
    CHECK(x2bar(2) == 10.0);
    CHECK(x2bar(3) == 11.0);
}

TEST_CASE("Grassmannian project_point", "[domain]")
{
    using Lielab::domain::Grassmannian;
    using Lielab::utils::to_VectorXd;
    
    const double pi = std::numbers::pi_v<double>;

    Eigen::MatrixXd axes(2, 1);
    axes(0,0) = std::cos(1.0/6.0*pi);
    axes(1,0) = std::sin(1.0/6.0*pi);

    const Grassmannian Gr12 = Grassmannian(to_VectorXd({0.0, 0.0}), axes);

    const Eigen::VectorXd vals1 = to_VectorXd({std::cos(1.0/6.0*pi), std::sin(1.0/6.0*pi)});
    const Eigen::VectorXd proj1 = Gr12.project_point(vals1);

    REQUIRE(proj1.size() == 2);
    CHECK_THAT(proj1(0), Catch::Matchers::WithinAbs(std::cos(1.0/6.0*pi), 1e-15));
    CHECK_THAT(proj1(1), Catch::Matchers::WithinAbs(std::sin(1.0/6.0*pi), 1e-15));

    const Eigen::VectorXd vals2 = to_VectorXd({-std::sin(1.0/6.0*pi), std::cos(1.0/6.0*pi)});
    const Eigen::VectorXd proj2 = Gr12.project_point(vals2);

    REQUIRE(proj2.size() == 2);
    CHECK_THAT(proj2(0), Catch::Matchers::WithinAbs(0.0, 1e-15));
    CHECK_THAT(proj2(1), Catch::Matchers::WithinAbs(0.0, 1e-15));

    const Eigen::VectorXd vals3 = to_VectorXd({std::cos(1.0/6.0*pi), std::sin(1.0/6.0*pi) - 4.0});
    const Eigen::VectorXd proj3 = Gr12.project_point(vals3);

    REQUIRE(proj3.size() == 2);
    CHECK_THAT(proj3(0), Catch::Matchers::WithinAbs(-std::cos(1.0/6.0*pi), 1e-15));
    CHECK_THAT(proj3(1), Catch::Matchers::WithinAbs(-std::sin(1.0/6.0*pi), 1e-15));
}

TEST_CASE("Grassmannian project_vector_onto_tangent_space", "[domain]")
{
    using Lielab::domain::Grassmannian;
    using Lielab::utils::to_VectorXd;
    
    const double pi = std::numbers::pi_v<double>;
    const Eigen::VectorXd vector = to_VectorXd({std::sqrt(2.0)/2.0, std::sqrt(2.0)/2.0});

    Eigen::MatrixXd axes1(2,1);
    axes1(0,0) = 1.0;
    axes1(1,0) = 0.0;
    const Grassmannian Gr121 = Grassmannian::project(to_VectorXd({0.0, 0.0}), axes1);
    const Eigen::VectorXd proj1 = Gr121.project_vector_onto_tangent_space(vector);

    REQUIRE(proj1.size() == 2);
    CHECK_THAT(proj1(0), Catch::Matchers::WithinAbs(std::sqrt(2.0)/2.0, 1e-15));
    CHECK_THAT(proj1(1), Catch::Matchers::WithinAbs(0.0, 1e-15));

    Eigen::MatrixXd axes2(2,1);
    axes2(0,0) = 0.0;
    axes2(1,0) = 1.0;
    const Grassmannian Gr122 = Grassmannian::project(to_VectorXd({0.0, 0.0}), axes2);
    const Eigen::VectorXd proj2 = Gr122.project_vector_onto_tangent_space(vector);

    REQUIRE(proj2.size() == 2);
    CHECK_THAT(proj2(0), Catch::Matchers::WithinAbs(0.0, 1e-15));
    CHECK_THAT(proj2(1), Catch::Matchers::WithinAbs(std::sqrt(2.0)/2.0, 1e-15));

    Eigen::MatrixXd axes3(2,1);
    axes3(0,0) = 1.0;
    axes3(1,0) = 1.0;
    const Grassmannian Gr123 = Grassmannian::project(to_VectorXd({0.0, 0.0}), axes3);
    const Eigen::VectorXd proj3 = Gr123.project_vector_onto_tangent_space(vector);

    REQUIRE(proj3.size() == 2);
    CHECK_THAT(proj3(0), Catch::Matchers::WithinAbs(std::sqrt(2.0)/2.0, 1e-15));
    CHECK_THAT(proj3(1), Catch::Matchers::WithinAbs(std::sqrt(2.0)/2.0, 1e-15));
}

TEST_CASE("Grassmannian project_vector_onto_normal_space", "[domain]")
{
    using Lielab::domain::Grassmannian;
    using Lielab::utils::to_VectorXd;
    
    const double pi = std::numbers::pi_v<double>;
    const Eigen::VectorXd vector = to_VectorXd({std::sqrt(2.0)/2.0, std::sqrt(2.0)/2.0});

    Eigen::MatrixXd axes1(2,1);
    axes1(0,0) = 1.0;
    axes1(1,0) = 0.0;
    const Grassmannian Gr121 = Grassmannian::project(to_VectorXd({0.0, 0.0}), axes1);
    const Eigen::VectorXd proj1 = Gr121.project_vector_onto_normal_space(vector);

    REQUIRE(proj1.size() == 2);
    CHECK_THAT(proj1(0), Catch::Matchers::WithinAbs(0.0, 1e-15));
    CHECK_THAT(proj1(1), Catch::Matchers::WithinAbs(std::sqrt(2.0)/2.0, 1e-15));

    Eigen::MatrixXd axes2(2,1);
    axes2(0,0) = 0.0;
    axes2(1,0) = 1.0;
    const Grassmannian Gr122 = Grassmannian::project(to_VectorXd({0.0, 0.0}), axes2);
    const Eigen::VectorXd proj2 = Gr122.project_vector_onto_normal_space(vector);

    REQUIRE(proj2.size() == 2);
    CHECK_THAT(proj2(0), Catch::Matchers::WithinAbs(std::sqrt(2.0)/2.0, 1e-15));
    CHECK_THAT(proj2(1), Catch::Matchers::WithinAbs(0.0, 1e-15));

    Eigen::MatrixXd axes3(2,1);
    axes3(0,0) = 1.0;
    axes3(1,0) = 1.0;
    const Grassmannian Gr123 = Grassmannian::project(to_VectorXd({0.0, 0.0}), axes3);
    const Eigen::VectorXd proj3 = Gr123.project_vector_onto_normal_space(vector);

    REQUIRE(proj3.size() == 2);
    CHECK_THAT(proj3(0), Catch::Matchers::WithinAbs(0.0, 1e-15));
    CHECK_THAT(proj3(1), Catch::Matchers::WithinAbs(0.0, 1e-15));
}

TEST_CASE("Grassmannian axes_intersection", "[domain]")
{
    using Lielab::domain::Grassmannian;
    using Lielab::utils::to_VectorXd;

    Eigen::MatrixXd axes1(2,1);
    axes1(0,0) = 1.0;
    axes1(1,0) = 0.0;
    const Grassmannian Gr121 = Grassmannian::project(to_VectorXd({0.0, 0.0}), axes1);
    const Eigen::VectorXd intersection1 = Gr121.axes_intersection(to_VectorXd({0.0, 1.0}), to_VectorXd({0.0, -1.0}));

    REQUIRE(intersection1.size() == 2);
    CHECK_THAT(intersection1(0), Catch::Matchers::WithinAbs(0.0, 1e-15));
    CHECK_THAT(intersection1(1), Catch::Matchers::WithinAbs(0.0, 1e-15));

    const Eigen::VectorXd intersection2 = Gr121.axes_intersection(to_VectorXd({0.0, 1.0}), to_VectorXd({1.0, 0.0}));

    REQUIRE(intersection2.size() == 2);
    CHECK(std::isnan(intersection2(0)));
    CHECK(std::isnan(intersection2(1)));

    const Eigen::VectorXd intersection3 = Gr121.axes_intersection(to_VectorXd({0.0, 1.0}), to_VectorXd({1.0, -1.0}));

    REQUIRE(intersection3.size() == 2);
    CHECK_THAT(intersection3(0), Catch::Matchers::WithinAbs(1.0, 1e-15));
    CHECK_THAT(intersection3(1), Catch::Matchers::WithinAbs(0.0, 1e-15));

    Eigen::MatrixXd axes2(2,1);
    axes2(0,0) = 1.0;
    axes2(1,0) = 1.0;
    const Grassmannian Gr122 = Grassmannian::project(to_VectorXd({0.0, 0.0}), axes2);
    const Eigen::VectorXd intersection4 = Gr122.axes_intersection(to_VectorXd({0.0, 1.0}), to_VectorXd({0.0, -1.0}));

    REQUIRE(intersection4.size() == 2);
    CHECK_THAT(intersection4(0), Catch::Matchers::WithinAbs(0.0, 1e-15));
    CHECK_THAT(intersection4(1), Catch::Matchers::WithinAbs(0.0, 1e-15));

    const Eigen::VectorXd intersection5 = Gr122.axes_intersection(to_VectorXd({0.0, 1.0}), to_VectorXd({1.0, 0.0}));

    REQUIRE(intersection5.size() == 2);
    CHECK_THAT(intersection5(0), Catch::Matchers::WithinAbs(1.0, 1e-15));
    CHECK_THAT(intersection5(1), Catch::Matchers::WithinAbs(1.0, 1e-15));

    const Eigen::VectorXd intersection6 = Gr122.axes_intersection(to_VectorXd({0.0, 1.0}), to_VectorXd({1.0, -1.0}));

    REQUIRE(intersection6.size() == 2);
    CHECK_THAT(intersection6(0), Catch::Matchers::WithinAbs(0.5, 1e-15));
    CHECK_THAT(intersection6(1), Catch::Matchers::WithinAbs(0.5, 1e-15));
}

File: python/tests/domain/smoothmanifolds/test_Grassmannian.py
import pytest

def test_Grassmannian_to_string():
    from lielab.domain import Grassmannian

    xblank = Grassmannian()
    assert (xblank.to_string() == "Grassmannian(0, 0, R)")

    x0 = Grassmannian(0, 0)
    assert (x0.to_string() == "Grassmannian(0, 0, R)")
    x1 = Grassmannian(1, 1)
    assert (x1.to_string() == "Grassmannian(1, 1, R)")
    x10 = Grassmannian(5, 10)
    assert (x10.to_string() == "Grassmannian(5, 10, R)")

def test_Grassmannian_main_initializer():
    from lielab.domain import Grassmannian

    xblank = Grassmannian()
    assert (xblank.get_dimension() == 0)

    x0 = Grassmannian(0, 0)
    assert (x0.get_dimension() == 0)
    x1 = Grassmannian(1, 1)
    assert (x1.get_dimension() == 1)
    x10 = Grassmannian(5, 10)
    assert (x10.get_dimension() == 5)

    with pytest.raises(RuntimeError):
        Grassmannian(6, 5)

def test_Grassmannian_get_dimension():
    from lielab.domain import Grassmannian

    zero = Grassmannian(0, 0)
    one = Grassmannian(1, 1)
    two = Grassmannian(2, 2)
    three = Grassmannian(3, 3)
    four = Grassmannian(4, 4)
    five = Grassmannian(5, 5)
    six = Grassmannian(6, 6)
    seven = Grassmannian(7, 7)
    eight = Grassmannian(8, 8)

    assert (zero.get_dimension() == 0)
    assert (one.get_dimension() == 1)
    assert (two.get_dimension() == 2)
    assert (three.get_dimension() == 3)
    assert (four.get_dimension() == 4)
    assert (five.get_dimension() == 5)
    assert (six.get_dimension() == 6)
    assert (seven.get_dimension() == 7)
    assert (eight.get_dimension() == 8)

def test_Grassmannian_serialize_unserialize():
    """
    Tests the serialize/unserialize operation.
    """

    from lielab.domain import Grassmannian

    x0 = Grassmannian(0, 0)
    x0.unserialize([])
    x0bar = x0.serialize()

    assert (x0bar.size == 0)

    x1 = Grassmannian(1, 2)
    x1.unserialize([1.0, 2.0, 3.0])
    x1bar = x1.serialize()

    assert (x1bar.size == 2)
    assert (x1bar[0] == 1.0)
    assert (x1bar[1] == 2.0)

    x1.unserialize([4.0, 5.0])
    x1bar = x1.serialize()

    assert (x1bar.size == 2)
    assert (x1bar[0] == 4.0)
    assert (x1bar[1] == 5.0)

    x1.unserialize([6.0])
    x1bar = x1.serialize()

    assert (x1bar.size == 2)
    assert (x1bar[0] == 6.0)
    assert (x1bar[1] == 5.0)

    x2 = Grassmannian(2, 4)
    x2.unserialize([1.0, 2.0, 3.0])
    x2bar = x2.serialize()

    assert (x2bar.size == 4)
    assert (x2bar[0] == 1.0)
    assert (x2bar[1] == 2.0)
    assert (x2bar[2] == 3.0)
    assert (x2bar[3] == 0.0)

    x2.unserialize([4.0, 5.0, 6.0, 7.0])
    x2bar = x2.serialize()

    assert (x2bar.size == 4)
    assert (x2bar[0] == 4.0)
    assert (x2bar[1] == 5.0)
    assert (x2bar[2] == 6.0)
    assert (x2bar[3] == 7.0)

    x2.unserialize([8.0, 9.0, 10.0, 11.0, 12.0])
    x2bar = x2.serialize()

    assert (x2bar.size == 4)
    assert (x2bar[0] == 8.0)
    assert (x2bar[1] == 9.0)
    assert (x2bar[2] == 10.0)
    assert (x2bar[3] == 11.0)

def test_Grassmannian_project_point():
    from lielab.domain import Grassmannian
    import numpy as np
    
    pi = np.pi
    axes = np.array([[np.cos(1.0/6.0*pi)], [np.sin(1.0/6.0*pi)]])

    Gr12 = Grassmannian([0.0, 0.0], axes)

    vals1 = np.array([np.cos(1.0/6.0*pi), np.sin(1.0/6.0*pi)])
    proj1 = Gr12.project_point(vals1)

    assert (proj1.size == 2)
    assert (np.abs(proj1[0] - np.cos(1.0/6.0*pi)) < 1e-15)
    assert (np.abs(proj1[1] - np.sin(1.0/6.0*pi)) < 1e-15)

    vals2 = np.array([-np.sin(1.0/6.0*pi), np.cos(1.0/6.0*pi)])
    proj2 = Gr12.project_point(vals2)

    assert (proj2.size == 2)
    assert (np.abs(proj2[0] - 0.0) < 1e-15)
    assert (np.abs(proj2[1] - 0.0) < 1e-15)

    vals3 = np.array([np.cos(1.0/6.0*pi), np.sin(1.0/6.0*pi) - 4.0])
    proj3 = Gr12.project_point(vals3)

    assert (proj3.size == 2)
    assert (np.abs(proj3[0] + np.cos(1.0/6.0*pi)) < 1e-15)
    assert (np.abs(proj3[1] + np.sin(1.0/6.0*pi)) < 1e-15)

def test_Grassmannian_project_vector_onto_tangent_space():
    from lielab.domain import Grassmannian
    import numpy as np
    
    pi = np.pi
    vector = np.array([np.sqrt(2.0)/2.0, np.sqrt(2.0)/2.0])

    axes1 = np.array([[1.0], [0.0]])
    Gr121 = Grassmannian.project([0.0, 0.0], axes1)
    proj1 = Gr121.project_vector_onto_tangent_space(vector)

    assert (proj1.size == 2)
    assert (np.abs(proj1[0] - np.sqrt(2.0)/2.0) < 1e-15)
    assert (np.abs(proj1[1] - 0.0) < 1e-15)

    axes2 = np.array([[0.0], [1.0]])
    Gr122 = Grassmannian.project([0.0, 0.0], axes2)
    proj2 = Gr122.project_vector_onto_tangent_space(vector)

    assert (proj2.size == 2)
    assert (np.abs(proj2[0] - 0.0) < 1e-15)
    assert (np.abs(proj2[1] - np.sqrt(2.0)/2.0) < 1e-15)

    axes3 = np.array([[1.0], [1.0]])
    Gr123 = Grassmannian.project([0.0, 0.0], axes3)
    proj3 = Gr123.project_vector_onto_tangent_space(vector)

    assert (proj3.size == 2)
    assert (np.abs(proj3[0] - np.sqrt(2.0)/2.0) < 1e-15)
    assert (np.abs(proj3[1] - np.sqrt(2.0)/2.0) < 1e-15)

def test_Grassmannian_project_vector_onto_normal_space():
    from lielab.domain import Grassmannian
    import numpy as np
    
    pi = np.pi
    vector = np.array([np.sqrt(2.0)/2.0, np.sqrt(2.0)/2.0])

    axes1 = np.array([[1.0], [0.0]])
    Gr121 = Grassmannian.project([0.0, 0.0], axes1)
    proj1 = Gr121.project_vector_onto_normal_space(vector)

    assert (proj1.size == 2)
    assert (np.abs(proj1[0] - 0.0) < 1e-15)
    assert (np.abs(proj1[1] - np.sqrt(2.0)/2.0) < 1e-15)

    axes2 = np.array([[0.0], [1.0]])
    Gr122 = Grassmannian.project([0.0, 0.0], axes2)
    proj2 = Gr122.project_vector_onto_normal_space(vector)

    assert (proj2.size == 2)
    assert (np.abs(proj2[0] - np.sqrt(2.0)/2.0) < 1e-15)
    assert (np.abs(proj2[1] - 0.0) < 1e-15)

    axes3 = np.array([[1.0], [1.0]])
    Gr123 = Grassmannian.project([0.0, 0.0], axes3)
    proj3 = Gr123.project_vector_onto_normal_space(vector)

    assert (proj3.size == 2)
    assert (np.abs(proj3[0] - 0.0) < 1e-15)
    assert (np.abs(proj3[1] - 0.0) < 1e-15)

def test_Grassmannian_axes_intersection():
    from lielab.domain import Grassmannian
    import numpy as np

    axes1 = np.array([[1.0], [0.0]])
    Gr121 = Grassmannian.project([0.0, 0.0], axes1)
    intersection1 = Gr121.axes_intersection([0.0, 1.0], [0.0, -1.0])

    assert (intersection1.size == 2)
    assert (np.abs(intersection1[0] - 0.0) < 1e-15)
    assert (np.abs(intersection1[1] - 0.0) < 1e-15)

    intersection2 = Gr121.axes_intersection([0.0, 1.0], [1.0, 0.0])

    assert (intersection2.size == 2)
    assert (np.isnan(intersection2[0]))
    assert (np.isnan(intersection2[1]))

    intersection3 = Gr121.axes_intersection([0.0, 1.0], [1.0, -1.0])

    assert (intersection3.size == 2)
    assert (np.abs(intersection3[0] - 1.0) < 1e-15)
    assert (np.abs(intersection3[1] - 0.0) < 1e-15)

    axes2 = np.array([[1.0], [1.0]])
    Gr122 = Grassmannian.project([0.0, 0.0], axes2)
    intersection4 = Gr122.axes_intersection([0.0, 1.0], [0.0, -1.0])

    assert (intersection4.size == 2)
    assert (np.abs(intersection4[0] - 0.0) < 1e-15)
    assert (np.abs(intersection4[1] - 0.0) < 1e-15)

    intersection5 = Gr122.axes_intersection([0.0, 1.0], [1.0, 0.0])

    assert (intersection5.size == 2)
    assert (np.abs(intersection5[0] - 1.0) < 1e-15)
    assert (np.abs(intersection5[1] - 1.0) < 1e-15)

    intersection6 = Gr122.axes_intersection([0.0, 1.0], [1.0, -1.0])

    assert (intersection6.size == 2)
    assert (np.abs(intersection6[0] - 0.5) < 1e-15)
    assert (np.abs(intersection6[1] - 0.5) < 1e-15)

